### ISSUE TITLE
Cannot use WatsonXChatLLM in an agent

### ISSUE THREAD

**Issue by nheidloff** [i-am-bee/bee-agent-framework#3]
I can run examples/llms/providers/watsonx.ts successfully. When I copy the code into examples/llms/agents/bee.ts and replace OllamaChatLLM I get the following error:

"Missing either space_id or project_id or wml_instance_crn"

```
yarn run start:bee
🛠️ System Please ensure that the code interpreter is running.
Interactive session has started. To escape, input 'q' and submit.
User 👤 : What's the current weather in Las Vegas?
ERR  🔥  11:40:13 (app): [AgentError: Error has occurred!] {
  isFatal: false,
  isRetryable: true,
  context: {},
  [errors]: [
    [RestfulClientError: Fetch has failed] {
      isFatal: false,
      isRetryable: false,
      context: {
        url: 'https://eu-de.ml.cloud.ibm.com/ml/v1/text/tokenization?version=2023-05-02',
        error: '{"errors":[{"code":"invalid_request_entity","message":"Missing either space_id or project_id or wml_instance_crn","more_info":"https://cloud.ibm.com/apidocs/watsonx-ai"}],"trace":"1f3b8c247c9e5200f4053a7383658971","status_code":400}',
        response: Response {
          status: 400,
          statusText: 'Bad Request',
          headers: Headers {
            date: 'Tue, 27 Aug 2024 11:40:13 GMT',
            'content-type': 'application/json',
            'content-length': '232',
            connection: 'keep-alive',
            'server-timing': 'intid;desc=a6da1fe00a08beab',
            'x-frame-options': 'DENY',
            'cache-control': 'no-cache, no-store, must-revalidate',
            'x-global-transaction-id': '1f3b8c247c9e5200f4053a7383658971',
            'content-security-policy': "default-src 'none'; script-src 'self'; connect-src 'self'; img-src 'self'; style-src 'self'; frame-ancestors 'none'; form-action 'self';",
            'strict-transport-security': 'max-age=31536000; includeSubDomains, max-age=31536000; includeSubDomains',
            'referrer-policy': 'strict-origin',
            pragma: 'no-cache',
            'x-xss-protection': '1; mode=block, 1; mode=block',
            'x-content-type-options': 'nosniff',
            'cf-cache-status': 'DYNAMIC',
            server: 'cloudflare',
            'cf-ray': '8b9bc91a2b7e162c-DUS'
          },
          body: ReadableStream {
            locked: true,
            state: 'closed',
            supportsBYOB: true
          },
          bodyUsed: true,
          ok: false,
          redirected: false,
          type: 'basic',
          url: 'https://eu-de.ml.cloud.ibm.com/ml/v1/text/tokenization?version=2023-05-02'
        }
      },
      [errors]: []
    }
  ]
}
```

I've changed the llm parameters to chatLLM:

```
const agent = new BeeAgent({
  llm: chatLLM,
  memory: new TokenMemory({ llm: chatLLM }),
  tools: [
    //new DuckDuckGoSearchTool(),
    new WebCrawlerTool(), // HTML web page crawler
    //new WikipediaTool(),
    new OpenMeteoTool(), // weather tool
    // new ArXivTool(), // research papers
    // new DynamicTool() // custom python tool
    ...(codeInterpreterUrl
      ? [
          new PythonTool({
            codeInterpreter: { url: codeInterpreterUrl },
            storage: new LocalPythonStorage({
              interpreterWorkingDir: `${__dirname}/tmp/code_interpreter`,
              localWorkingDir: `${__dirname}/tmp/local`,
            }),
          }),
        ]
      : []),
  ],
});
```

I just cloned the repo and use main. Am I doing anything wrong?
Thanks!

**Comment by Tomas2D** [comment]
Hello. Thanks for the report; 

The `tokenize` method in the `WatsonXLLM` class was not sending `project_id`/`space_id` to the appropriate WatsonX AI API, and that caused the error.

Because WatsonX does not have a chat endpoint, you must initiate the LLM like this to work with the agent.

```typescript
const template = new PromptTemplate({
  variables: ["messages"],
  template: `{{#messages}}{{#system}}<|begin_of_text|><|start_header_id|>system<|end_header_id|>

{{system}}<|eot_id|>{{/system}}{{#user}}<|start_header_id|>user<|end_header_id|>

{{user}}<|eot_id|>{{/user}}{{#assistant}}<|start_header_id|>assistant<|end_header_id|>

{{assistant}}<|eot_id|>{{/assistant}}{{/messages}}<|start_header_id|>assistant<|end_header_id|>

`,
});

const llm = new WatsonXChatLLM({
  llm: new WatsonXLLM({
    modelId: "meta-llama/llama-3-70b-instruct",
    projectId: process.env.WATSONX_PROJECT_ID,
    apiKey: process.env.WATSONX_API_KEY,
    parameters: {
      decoding_method: "greedy",
      max_new_tokens: 50,
    },
    deploymentId: process.env.WATSONX_DEPLOYMENT_ID
  }),
  config: {
    messagesToPrompt(messages: BaseMessage[]) {
      return template.render({
        messages: messages.map((message) => ({
          system: message.role === "system" ? [message.text] : [],
          user: message.role === "user" ? [message.text] : [],
          assistant: message.role === "assistant" ? [message.text] : [],
        })),
      });
    },
  },
});
``` 

Try the new version, 0.0.2, and let me know.

**Comment by nheidloff** [comment]
Thanks @Tomas2D for the quick fix! It works now.

**Comment by LordHeImchen** [comment]
Hi @Tomas2D,
is it possible that the solution provided above is deprecated? 
When running this code I am experiencing an issue in the upper part of the code, where the template gets specified. Specifically, it tells me that:

```
Object literal may only specify known properties, and 'variables' does not exist in type 'Omit<PromptTemplateInput<ZodType<any, ZodTypeDef, any>>, "schema"> & { schema: ZodType<any, ZodTypeDef, any> | SchemaObject; }'.
```

So what I tried to do, was solving it by specifying it as follows:
```typescript
const llm = new WatsonXLLM({
    modelId: "meta-llama/llama-3-70b-instruct",
    projectId: process.env.WATSONX_PROJECT_ID,
    apiKey: process.env.WATSONX_API_KEY,
    baseUrl: process.env.WATSONX_BASE_URL,
    // deploymentId: process.env.WATSONX_DEPLOYMENT_ID,
    parameters: {
      decoding_method: "greedy",
      max_new_tokens: 250,
    },
});

const template = new PromptTemplate({
    schema: {
        messages: {
            "system": "",
            "user": "",
            "assistant": "",
        },
    },
    template: `{{#messages}}{{#system}}<|begin_of_text|><|start_header_id|>system<|end_header_id|>

{{system}}<|eot_id|>{{/system}}{{#user}}<|start_header_id|>user<|end_header_id|>

{{user}}<|eot_id|>{{/user}}{{#assistant}}<|start_header_id|>assistant<|end_header_id|>

{{assistant}}<|eot_id|>{{/assistant}}{{/messages}}<|start_header_id|>assistant<|end_header_id|>

`,
});

const chatLLM = new WatsonXChatLLM({
    llm,
    config: {
      messagesToPrompt(messages: BaseMessage[]) {
        return template.render({
          messages: messages.map((message) => ({
            system: message.role === "system" ? [message.text] : [],
            user: message.role === "user" ? [message.text] : [],
            assistant: message.role === "assistant" ? [message.text] : [],
          })),
        });
      },
    },
  });
``` 

This works fine, using the `config` you provided above for the WatsonXChatLLM. However, running the following agent will always throw an error as soon as it tries to leverage a tool.

```typescript
const agent = new BeeAgent({
    llm: chatLLM,
    memory: new TokenMemory({ llm: chatLLM }),
    tools: [
      new OpenMeteoTool()
    ],
});

try {
  for await (const { prompt } of reader) {
    const response = await agent.run(
        { prompt },
        {
          execution: {
            maxRetriesPerStep: 3,
            totalMaxRetries: 10,
            maxIterations: 20,
          },
        },
      ).observe((emitter) => {
          emitter.on("start", () => {
            reader.write(`Agent 🤖 : `, "starting new iteration");
          });
          emitter.on("error", ({ error }) => {
            reader.write(`Agent 🤖 : `, FrameworkError.ensure(error).dump());
          });
          emitter.on("retry", () => {
            reader.write(`Agent 🤖 : `, "retrying the action...");
          });
          emitter.on("update", async ({ data, update, meta }) => {
            reader.write(`Agent (${update.key}) 🤖 : `, update.value);
          });
          emitter.match("*.*", async (data: any, event) => {
            if (event.creator === chatLLM) {
              const eventName = event.name as keyof GenerateCallbacks;
              switch (eventName) {
                case "start":
                  console.info("LLM Input");
                  console.info(data.input);
                  break;
                case "success":
                  console.info("LLM Output");
                  console.info(data)
                  // console.info(data.value.raw.finalResult);
                  break;
                case "error":
                  console.error(data);
                  break;
              }
            }
          });
      });
    reader.write(`Agent 🤖 : `, response.result.text);
  }
} catch (error) {
  logger.error(FrameworkError.ensure(error).dump());
} finally {
  process.exit(0);
}
```

As I said above, it works as long as I ask for some standard information. As soon as I prompt it for the weather I receive this error:
```
Agent 🤖 :  [LLMError: LLM has occurred an error.] {
  isFatal: false,
  isRetryable: true,
  context: {},
  [errors]: [
    [LinePrefixParserError: The generated output does not adhere to the schema.
    Parsed text line corresponds to a node "tool_name" which is not a start node!] {
      isFatal: true,
      isRetryable: false,
      context: {
        lines: '\nFunction Name:',
        excludedLines: ': The user wants to know the current weather in Paris. I can use the OpenMeteo function to retrieve the weather forecast.',
        finalState: {},
        partialState: {}
      },
      [errors]: []
    }
  ]
}
```

So basically my questions are:
- How can I make `tool_name` a start node? I had a look at the other examples but couldn't find any major differences...
- Also, assume I would want to use the model `mistralai/mistral-large`, where could I find the specific template which needs to be used for this model? (I could only find three examples under src/adapters/shared/llmChatTemplate.ts)

Any help is appreciated, thanks for this cool framework!
